### ISSUE TITLE
BUG: 

### ISSUE THREAD

**Issue by seanslma** [pandas-dev/pandas#55212]
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
t1 = '2023-09-01'
t2 = '2023-09-01 01:00:00'
t3 = '2023-09-01 01:30:00'
ds1 = pd.date_range(t1, t2, freq='30T')
ds2 = pd.date_range(t1, t3, freq='30T')
df1 = pd.DataFrame({
    'ds': ds1.astype('datetime64[us]'),
    'y1': range(len(ds1)),
})
df2 = pd.DataFrame({
    'ds': ds2.astype('datetime64[ns]'),
    'y2': range(len(ds2)),
})
df3 = df1.merge(df2, on=['ds'], how='outer') #will only convert df2 `ds` type to df1 `ds` type
df3.dtypes
```

### Issue Description

The column type was not converted from the lower resolution (dtype('<M8[us]')) to the higher resolution (dtype('<M8[ns]')). 

It only converted the type in df2 to the type in df2 for the `on` columns. 
```
File ~\conda-envs\dev\lib\site-packages\pandas\core\arrays\_mixins.py:398, in NDArrayBackedExtensionArray._putmask(self, mask, value)
    383 """
    384 Analogue to np.putmask(self, mask, value)
    385 
   (...)
    394     If value cannot be cast to self.dtype.
    395 """
    396 value = self._validate_setitem_value(value)
--> 398 np.putmask(self._ndarray, mask, value)

File <__array_function__ internals>:180, in putmask(*args, **kwargs)

TypeError: Cannot cast array data from dtype('<M8[ns]') to dtype('<M8[us]') according to the rule 'safe'
```

### Expected Behavior

I expect `df1.merge(df2, on=['ds'], how='outer')` and `df2.merge(df1, on=['ds'], how='outer')` both should work in this case.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit              : ba1cccd19da778f0c3a7d6a885685da16a072870
python              : 3.9.15.final.0
python-bits         : 64
OS                  : Windows
OS-release          : 10
Version             : 10.0.19044
machine             : AMD64
processor           : Intel64 Family 6 Model 58 Stepping 0, GenuineIntel
byteorder           : little
LC_ALL              : None
LANG                : None
LOCALE              : English_Australia.1252

pandas              : 2.1.0
numpy               : 1.23.3
pytz                : 2022.1
dateutil            : 2.8.2
setuptools          : 65.5.0
pip                 : 22.3.1
Cython              : 0.29.30
pytest              : 7.1.2
hypothesis          : None
sphinx              : None
blosc               : None
feather             : None
xlsxwriter          : None
lxml.etree          : 4.9.0
html5lib            : None
pymysql             : None
psycopg2            : 2.9.3
jinja2              : 3.0.3
IPython             : 8.4.0
pandas_datareader   : None
bs4                 : 4.11.1
bottleneck          : 1.3.5
dataframe-api-compat: None
fastparquet         : None
fsspec              : 2022.5.0
gcsfs               : None
matplotlib          : 3.5.2
numba               : 0.53.0
numexpr             : 2.7.3
odfpy               : None
openpyxl            : 3.0.9
pandas_gbq          : None
pyarrow             : 12.0.0
pyreadstat          : None
pyxlsb              : None
s3fs                : None
scipy               : 1.8.1
sqlalchemy          : 1.4.46
tables              : None
tabulate            : None
xarray              : 2022.3.0
xlrd                : None
zstandard           : None
tzdata              : 2022.1
qtpy                : None
pyqt5               : None

</details>

**Comment by paulreece** [comment]
I have confirmed locally that this bug occurs on the main development branch.

**Comment by rhshadrach** [comment]
@mroeschke @MarcoGorelli any thoughts on what should happen here?

**Comment by mroeschke** [comment]
Yeah as mentioned in the OP this should work by casting to the higher resolution. Concat has as similar problem https://github.com/pandas-dev/pandas/pull/53641#discussion_r1273869349

**Comment by lithomas1** [comment]
I don't think its a regression, I get the same error on 2.0.3, too.

Should have a PR up for 2.1.4, anyways, though.
